Lazily allocate ResourceResponse within CachedResource

mattwoodrow opened the pull request:
#### 98761d6864f7dac2f22655a51bb8a00a2172edc3
<pre>
Lazily allocate ResourceResponse within CachedResource
<a href="https://bugs.webkit.org/show_bug.cgi?id=252637">https://bugs.webkit.org/show_bug.cgi?id=252637</a>

Reviewed by Cameron McCormack.

CachedResource objects often don&apos;t have an initialised ResourceResponse, especially in the case of Fonts.

We can move the ResourceResponse (and associated response data) into a lazily allocated inner structure to save memory

* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::checkNotify):
(WebCore::CachedCSSStyleSheet::responseMIMEType const):
(WebCore::CachedCSSStyleSheet::mimeTypeAllowedByNosniff const):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::CachedImage):
(WebCore::CachedImage::isPDFResource const):
(WebCore::CachedImage::isPostScriptResource const):
(WebCore::CachedImage::responseReceived):
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::didAddClient):
(WebCore::CachedRawResource::responseReceived):
(WebCore::CachedRawResource::previewResponseReceived):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource):
(WebCore::CachedResource::setBodyDataFrom):
(WebCore::CachedResource::cancelLoad):
(WebCore::CachedResource::isExpired const):
(WebCore::CachedResource::setResponse):
(WebCore::CachedResource::didAddClient):
(WebCore::CachedResource::addClientToSet):
(WebCore::CachedResource::destroyDecodedDataIfNeeded):
(WebCore::CachedResource::setDecodedSize):
(WebCore::CachedResource::setEncodedSize):
(WebCore::CachedResource::updateResponseAfterRevalidation):
(WebCore::CachedResource::canUseCacheValidator const):
(WebCore::CachedResource::makeRevalidationDecision const):
(WebCore::CachedResource::overheadSize const):
(WebCore::CachedResource::ResponseData::ResponseData):
(WebCore::CachedResource::mutableResponseData const):
(WebCore::CachedResource::mutableResponse):
(WebCore::CachedResource::response const):
(WebCore::CachedResource::stopDecodedDataDeletionTimer):
(WebCore::CachedResource::restartDecodedDataDeletionTimer):
(WebCore::CachedResource::resourceError const):
(WebCore::CachedResource::wasCanceled const):
(WebCore::CachedResource::loadFailedOrCanceled const):
(WebCore::CachedResource::encodedSize const):
(WebCore::CachedResource::decodedSize const):
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::setResourceError):
(WebCore::CachedResource::mimeType const):
(WebCore::CachedResource::expectedContentLength const):
(WebCore::CachedResource::takeNetworkLoadMetrics):
(WebCore::CachedResource::resourceError const): Deleted.
(WebCore::CachedResource::encodedSize const): Deleted.
(WebCore::CachedResource::decodedSize const): Deleted.
(WebCore::CachedResource::response const): Deleted.
(WebCore::CachedResource::wasCanceled const): Deleted.
(WebCore::CachedResource::loadFailedOrCanceled const): Deleted.
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::didAddClient):
(WebCore::CachedXSLStyleSheet::checkNotify):

Canonical link: <a href="https://commits.webkit.org/260788@main">https://commits.webkit.org/260788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f338295282d8d804388baf7ce1c06e06370ecb3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/935 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113322 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9758 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101712 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115193 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84872 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11305 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9328 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11969 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/50744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7467 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13700 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -905,7 +903,7 @@ bool CachedResource::varyHeaderValuesMatch(const ResourceRequest& request)
 {
     static const int kAverageClientsHashMapSize = 384;
-    return sizeof(CachedResource) + m_response.memoryUsage() + kAverageClientsHashMapSize + m_resourceRequest.url().string().length() * 2;
+    return sizeof(CachedResource) + response().memoryUsage() + kAverageClientsHashMapSize + m_resourceRequest.url().string().length() * 2;
```
Should we only count `response().memoryUsage()` if it's not the static one?

mattwoodrow replied:
Unfortunately we can't, since the MemoryCache uses this result as a key, and it can't change without corrupting the cache index.

heycam commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -916,6 +914,87 @@ void CachedResource::setLoadPriority(const std::optional<ResourceLoadPriority>&
+const ResourceError& CachedResource::resourceError() const
+{
+    if (!m_response) {
+        static LazyNeverDestroyed<ResourceError> emptyError;
+        static std::once_flag onceFlag;
+        std::call_once(onceFlag, [&] {
+            emptyError.construct();
+        });
+        return emptyError;
```
I don't think we need this, since the ResourceError lives in the ResponseData, and we can just return the one from response().

mattwoodrow replied:
response() is a ResourceResponse, not ResponseData.

heycam commented on Source/WebCore/loader/cache/CachedResource.h:
```diff
@@ -345,6 +347,24 @@ class CachedResource {
+
+    struct ResponseData {
+        WTF_MAKE_NONCOPYABLE(ResponseData);
+        WTF_MAKE_FAST_ALLOCATED;
```
Can you use WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER?

mattwoodrow committed:
Lazily allocate ResourceResponse within CachedResource
https://bugs.webkit.org/show_bug.cgi?id=252637

Reviewed by Cameron McCormack.

CachedResource objects often don't have an initialised ResourceResponse, especially in the case of Fonts.

We can move the ResourceResponse (and associated response data) into a lazily allocated inner structure to save memory

* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::checkNotify):
(WebCore::CachedCSSStyleSheet::responseMIMEType const):
(WebCore::CachedCSSStyleSheet::mimeTypeAllowedByNosniff const):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::CachedImage):
(WebCore::CachedImage::isPDFResource const):
(WebCore::CachedImage::isPostScriptResource const):
(WebCore::CachedImage::responseReceived):
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::didAddClient):
(WebCore::CachedRawResource::responseReceived):
(WebCore::CachedRawResource::previewResponseReceived):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource):
(WebCore::CachedResource::setBodyDataFrom):
(WebCore::CachedResource::cancelLoad):
(WebCore::CachedResource::isExpired const):
(WebCore::CachedResource::setResponse):
(WebCore::CachedResource::didAddClient):
(WebCore::CachedResource::addClientToSet):
(WebCore::CachedResource::destroyDecodedDataIfNeeded):
(WebCore::CachedResource::setDecodedSize):
(WebCore::CachedResource::setEncodedSize):
(WebCore::CachedResource::updateResponseAfterRevalidation):
(WebCore::CachedResource::canUseCacheValidator const):
(WebCore::CachedResource::makeRevalidationDecision const):
(WebCore::CachedResource::overheadSize const):
(WebCore::CachedResource::ResponseData::ResponseData):
(WebCore::CachedResource::mutableResponseData const):
(WebCore::CachedResource::mutableResponse):
(WebCore::CachedResource::response const):
(WebCore::CachedResource::stopDecodedDataDeletionTimer):
(WebCore::CachedResource::restartDecodedDataDeletionTimer):
(WebCore::CachedResource::resourceError const):
(WebCore::CachedResource::wasCanceled const):
(WebCore::CachedResource::loadFailedOrCanceled const):
(WebCore::CachedResource::encodedSize const):
(WebCore::CachedResource::decodedSize const):
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::setResourceError):
(WebCore::CachedResource::mimeType const):
(WebCore::CachedResource::expectedContentLength const):
(WebCore::CachedResource::takeNetworkLoadMetrics):
(WebCore::CachedResource::resourceError const): Deleted.
(WebCore::CachedResource::encodedSize const): Deleted.
(WebCore::CachedResource::decodedSize const): Deleted.
(WebCore::CachedResource::response const): Deleted.
(WebCore::CachedResource::wasCanceled const): Deleted.
(WebCore::CachedResource::loadFailedOrCanceled const): Deleted.
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::didAddClient):
(WebCore::CachedXSLStyleSheet::checkNotify):

Canonical link: https://commits.webkit.org/260788@main
